Build difficulties on CentOS 8

I was able to build `ebpfpub` on CentOS 8 but there were some obstacles. I want to make sure I did this right, and if so, maybe we can update the `README` regarding the needed dependencies.

1. I downloaded the osquery toolchain and gave the path to it, but the configure step fails with:

```
/home/trailofbits/Downloads/osquery-toolchain-1.1.0-x86_64/usr/bin/clang is not a full path to an existing compiler tool.
```

Workaround: build with the system toolchain instead.

2. proceeding to build with the system toolchain, fails with:

```
could not find a package configuration file provided by "LLVM" with any of the following names
```

Solution: `yum install llvm-devel` which installs the missing the `LLVMConfig.cmake` files.

3. Linking errors because llvm-dev is missing a bunch of libraries:

```
/usr/bin/ld: cannot find -lLLVMCore 
/usr/bin/ld: cannot find -lLLVMBPFCodeGen
 /usr/bin/ld: cannot find -lLLVMX86CodeGen 
/usr/bin/ld: cannot find -lLLVMExecutionEngine 
/usr/bin/ld: cannot find -lLLVMMCJIT
```

Solution: `yum install llvm-static` to install the missing libraries.

4. One last linking error: `/usr/bin/ld: cannot find -ltinfo`

Solution: `yum install ncurses-devel` (installs the `libtinfo` library dependency)

--- a/README.md
+++ b/README.md
@@ -37,11 +37,11 @@ tar -xf /tmp/ebpfpub/build/osquery-toolchain-1.1.0-x86_64.tar.xz -C /opt
 ### Steps to Build
 
 1. Obtain the source code: `git clone --recursive https://github.com/trailofbits/ebpfpub`
-2. In case the `--recursive` flag was not provided, run `git submodule update --init --recursive`
+2. If you cloned the repo without the `--recursive` flag, run `git submodule update --init --recursive`
 3. Enter the source folder: `cd ebpfpub`
-4. Create the build folder: `mkdir build && cd build`
-5. Configure the project: `cmake -DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo -DEBPF_COMMON_TOOLCHAIN_PATH:PATH=/path/to/osquery-toolchain -DEBPFPUB_ENABLE_INSTALL:BOOL=true -DEBPFPUB_ENABLE_EXAMPLES:BOOL=true -DEBPF_COMMON_ENABLE_TESTS:BOOL=true ..`
-(remove `-DEBPF_COMMON_TOOLCHAIN_PATH:PATH=/path/to/osquery-toolchain` if you are building with the system toolchain)
+4. Create and enter the build folder: `mkdir build && cd build`
+5. Configure the project: `cmake -DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo -DEBPF_COMMON_TOOLCHAIN_PATH:PATH=/opt/osquery-toolchain -DEBPFPUB_ENABLE_INSTALL:BOOL=true -DEBPFPUB_ENABLE_EXAMPLES:BOOL=true -DEBPF_COMMON_ENABLE_TESTS:BOOL=true ..`
+(remove `-DEBPF_COMMON_TOOLCHAIN_PATH:PATH=/opt/osquery-toolchain` if you are building with the system toolchain)
 6. Build the project: `cmake --build . -j $(($(nproc) + 1))`
 7. Run the tests: `cmake --build . --target run-ebpf-common-tests`
 

--- a/README.md
+++ b/README.md
@@ -9,10 +9,12 @@ ebpfpub is a generic function tracing library for Linux that supports tracepoint
 ## Building
 
 ### Prerequisites
-* A recent Clang/LLVM installation (8.0 or better), compiled with BPF support
+* A recent Clang/LLVM installation (8.0 or better), compiled with BPF support.
+  * Test for the support: `llc --version | grep bpf` and check that BPF is listed as a registered target.
 * A recent libc++ or stdc++ library, supporting C++17
 * CMake >= 3.16.2. A pre-built binary can be downloaded from the [CMake's download page](https://cmake.org/download/).
-* Linux kernel >= 4.18 (Ubuntu 18.10)
+* Linux kernel >= 4.18 (Ubuntu 18.10, CentOS 8, Red Hat Enterprise Linux 8).
+  * Test for the support: ``grep BPF /boot/config-`uname -r` `` and check the output for `CONFIG_BPF=y` and `CONFIG_BPF_SYSCALL=y`
 
 Please note that LLVM itself must be compiled with libc++ when enabling the `EBPF_COMMON_ENABLE_LIBCPP` option, since ebfpub will directly link against the LLVM libraries.
 

--- a/README.md
+++ b/README.md
@@ -13,6 +13,7 @@ ebpfpub is a generic function tracing library for Linux that supports tracepoint
 * CMake >= 3.16.2. A pre-built binary can be downloaded from the [CMake's download page](https://cmake.org/download/).
 * Linux kernel >= 4.18 (Ubuntu 18.10, CentOS 8, Red Hat Enterprise Linux 8).
   * Test for the support: ``grep BPF /boot/config-`uname -r` `` and check the output for `CONFIG_BPF=y` and `CONFIG_BPF_SYSCALL=y`
+* The package `libz-dev`, needed during linking.
 * Optional, but highly recommended: Download the osquery-toolchain: https://github.com/osquery/osquery-toolchain/releases/download/1.1.0/osquery-toolchain-1.1.0-x86_64.tar.xz
   * **This should work fine on any recent Linux distribution. The binaries generated with this toolchain are portable and can be deployed on any distro >= CentOS 6/Ubuntu 16.04**
 * If _not_ using the osquery-toolchain (if building with the system toolchain):

--- a/README.md
+++ b/README.md
@@ -9,6 +9,7 @@ ebpfpub is a generic function tracing library for Linux that supports tracepoint
 ## Building
 
 ### Prerequisites
+
 * A recent libc++ or stdc++ library, supporting C++17
 * CMake >= 3.16.2. A pre-built binary can be downloaded from the [CMake's download page](https://cmake.org/download/).
 * Linux kernel >= 4.18 (Ubuntu 18.10, CentOS 8, Red Hat Enterprise Linux 8).
@@ -24,14 +25,16 @@ ebpfpub is a generic function tracing library for Linux that supports tracepoint
   * The packages `llvm-devel` (for `LLVMConfig.cmake` files), `llvm-static` (for additional LLVM libraries), and `ncurses-devel` (for `libtinfo`)
 
 #### Installing the osquery-toolchain
+
 As root:
-```bash
+```shell
 cd /tmp
 wget https://github.com/osquery/osquery-toolchain/releases/download/1.1.0/osquery-toolchain-1.1.0-x86_64.tar.xz 
 tar -xf /tmp/ebpfpub/build/osquery-toolchain-1.1.0-x86_64.tar.xz -C /opt
 ```
 
 ### Dependencies (retrieved with git)
+
 * [ebpf-common](https://github.com/trailofbits/ebpf-common)
 
 ### Steps to Build
@@ -45,17 +48,19 @@ tar -xf /tmp/ebpfpub/build/osquery-toolchain-1.1.0-x86_64.tar.xz -C /opt
 6. Build the project: `cmake --build . -j $(($(nproc) + 1))`
 7. Run the tests: `cmake --build . --target run-ebpf-common-tests`
 
-### Building the packages
+## Building the package
 
-## Prerequisites
+### Prerequisites for packaging
+
 * DEB: **dpkg** command
 * RPM: **rpm** command
 * TGZ: **tar** command
 
-## Steps
+### Steps to package
+
 Make sure that the `-DEBPFPUB_ENABLE_INSTALL:BOOL=true` parameter has been passed at configure time, then run the following commands inside the build folder:
 
-```
+```shell
 mkdir install
 export DESTDIR=`realpath install`
 
@@ -65,7 +70,7 @@ cmake --build . --target install
 
 Configure the packaging project:
 
-```
+```shell
 mkdir package
 cd package
 

--- a/README.md
+++ b/README.md
@@ -9,23 +9,23 @@ ebpfpub is a generic function tracing library for Linux that supports tracepoint
 ## Building
 
 ### Prerequisites
-* A recent Clang/LLVM installation (8.0 or better), compiled with BPF support.
-  * Test for the support: `llc --version | grep bpf` and check that BPF is listed as a registered target.
 * A recent libc++ or stdc++ library, supporting C++17
 * CMake >= 3.16.2. A pre-built binary can be downloaded from the [CMake's download page](https://cmake.org/download/).
 * Linux kernel >= 4.18 (Ubuntu 18.10, CentOS 8, Red Hat Enterprise Linux 8).
   * Test for the support: ``grep BPF /boot/config-`uname -r` `` and check the output for `CONFIG_BPF=y` and `CONFIG_BPF_SYSCALL=y`
+* Optional, but highly recommended: Download the osquery-toolchain: https://github.com/osquery/osquery-toolchain/releases/download/1.1.0/osquery-toolchain-1.1.0-x86_64.tar.xz
+* If _not_ using the osquery-toolchain (if building with the system toolchain):
+  * A recent Clang/LLVM installation (8.0 or better), compiled with BPF support.
+    * Test for the support: `llc --version | grep bpf` and check that BPF is listed as a registered target.
+    * Please note that LLVM itself must be compiled with libc++ when enabling the `EBPF_COMMON_ENABLE_LIBCPP` option, since ebfpub will directly link against the LLVM libraries.
+  * The packages `llvm-devel` (for `LLVMConfig.cmake` files), `llvm-static` (for additional LLVM libraries), and `ncurses-devel` (for `libtinfo`)
 
-Please note that LLVM itself must be compiled with libc++ when enabling the `EBPF_COMMON_ENABLE_LIBCPP` option, since ebfpub will directly link against the LLVM libraries.
-
-### Dependencies
+### Dependencies (retrieved with git)
 * [ebpf-common](https://github.com/trailofbits/ebpf-common)
 
 ### Building with the osquery toolchain (preferred)
 
 **This should work fine on any recent Linux distribution. The binaries generated with this toolchain are portable and can be deployed on any distro >= CentOS 6/Ubuntu 16.04**
-
-Download the osquery-toolchain: https://github.com/osquery/osquery-toolchain/releases/download/1.1.0/osquery-toolchain-1.1.0-x86_64.tar.xz
 
 1. Obtain the source code: `git clone --recursive https://github.com/trailofbits/ebpfpub`
 2. In case the `--recursive` flag was not provided, run `git submodule update --init --recursive`
@@ -37,7 +37,7 @@ Download the osquery-toolchain: https://github.com/osquery/osquery-toolchain/rel
 
 ### Building with the system toolchain
 
-**Note that this will fail unless clang and the C++ library both support C++17**. Recent distributions should be compatible (tested on Arch Linux, Ubuntu 19.10).
+**Note that this will fail unless Clang and the C++ library both support C++17**. Recent distributions should be compatible (tested on Arch Linux, Ubuntu 19.10 and above).
 
 1. Obtain the source code: `git clone --recursive https://github.com/trailofbits/ebpfpub`
 2. In case the `--recursive` flag was not provided, run `git submodule update --init --recursive`

--- a/README.md
+++ b/README.md
@@ -14,7 +14,9 @@ ebpfpub is a generic function tracing library for Linux that supports tracepoint
 * Linux kernel >= 4.18 (Ubuntu 18.10, CentOS 8, Red Hat Enterprise Linux 8).
   * Test for the support: ``grep BPF /boot/config-`uname -r` `` and check the output for `CONFIG_BPF=y` and `CONFIG_BPF_SYSCALL=y`
 * Optional, but highly recommended: Download the osquery-toolchain: https://github.com/osquery/osquery-toolchain/releases/download/1.1.0/osquery-toolchain-1.1.0-x86_64.tar.xz
+  * **This should work fine on any recent Linux distribution. The binaries generated with this toolchain are portable and can be deployed on any distro >= CentOS 6/Ubuntu 16.04**
 * If _not_ using the osquery-toolchain (if building with the system toolchain):
+  * **Clang and the C++ library must both support C++17**. Recent distributions should be compatible (tested on Arch Linux, Ubuntu 19.10 and above).
   * A recent Clang/LLVM installation (8.0 or better), compiled with BPF support.
     * Test for the support: `llc --version | grep bpf` and check that BPF is listed as a registered target.
     * Please note that LLVM itself must be compiled with libc++ when enabling the `EBPF_COMMON_ENABLE_LIBCPP` option, since ebfpub will directly link against the LLVM libraries.
@@ -23,27 +25,14 @@ ebpfpub is a generic function tracing library for Linux that supports tracepoint
 ### Dependencies (retrieved with git)
 * [ebpf-common](https://github.com/trailofbits/ebpf-common)
 
-### Building with the osquery toolchain (preferred)
-
-**This should work fine on any recent Linux distribution. The binaries generated with this toolchain are portable and can be deployed on any distro >= CentOS 6/Ubuntu 16.04**
+### Steps to Build
 
 1. Obtain the source code: `git clone --recursive https://github.com/trailofbits/ebpfpub`
 2. In case the `--recursive` flag was not provided, run `git submodule update --init --recursive`
 3. Enter the source folder: `cd ebpfpub`
 4. Create the build folder: `mkdir build && cd build`
 5. Configure the project: `cmake -DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo -DEBPF_COMMON_TOOLCHAIN_PATH:PATH=/path/to/osquery-toolchain -DEBPFPUB_ENABLE_INSTALL:BOOL=true -DEBPFPUB_ENABLE_EXAMPLES:BOOL=true -DEBPF_COMMON_ENABLE_TESTS:BOOL=true ..`
-6. Build the project: `cmake --build . -j $(($(nproc) + 1))`
-7. Run the tests: `cmake --build . --target run-ebpf-common-tests`
-
-### Building with the system toolchain
-
-**Note that this will fail unless Clang and the C++ library both support C++17**. Recent distributions should be compatible (tested on Arch Linux, Ubuntu 19.10 and above).
-
-1. Obtain the source code: `git clone --recursive https://github.com/trailofbits/ebpfpub`
-2. In case the `--recursive` flag was not provided, run `git submodule update --init --recursive`
-3. Enter the source folder: `cd ebpfpub`
-4. Create the build folder: `mkdir build && cd build`
-5. Configure the project: `cmake -DCMAKE_BUILD_TYPE:STRING=RelWithDebInfo -DCMAKE_C_COMPILER:STRING=clang -DCMAKE_CXX_COMPILER=clang++ -DEBPFPUB_ENABLE_INSTALL:BOOL=true -DEBPFPUB_ENABLE_EXAMPLES:BOOL=true -DEBPF_COMMON_ENABLE_TESTS:BOOL=true ..`
+(remove `-DEBPF_COMMON_TOOLCHAIN_PATH:PATH=/path/to/osquery-toolchain` if you are building with the system toolchain)
 6. Build the project: `cmake --build . -j $(($(nproc) + 1))`
 7. Run the tests: `cmake --build . --target run-ebpf-common-tests`
 

--- a/README.md
+++ b/README.md
@@ -14,7 +14,7 @@ ebpfpub is a generic function tracing library for Linux that supports tracepoint
 * Linux kernel >= 4.18 (Ubuntu 18.10, CentOS 8, Red Hat Enterprise Linux 8).
   * Test for the support: ``grep BPF /boot/config-`uname -r` `` and check the output for `CONFIG_BPF=y` and `CONFIG_BPF_SYSCALL=y`
 * The package `libz-dev`, needed during linking.
-* Optional, but highly recommended: Download the osquery-toolchain: https://github.com/osquery/osquery-toolchain/releases/download/1.1.0/osquery-toolchain-1.1.0-x86_64.tar.xz
+* Optional, but highly recommended: download and install the osquery-toolchain (see below).
   * **This should work fine on any recent Linux distribution. The binaries generated with this toolchain are portable and can be deployed on any distro >= CentOS 6/Ubuntu 16.04**
 * If _not_ using the osquery-toolchain (if building with the system toolchain):
   * **Clang and the C++ library must both support C++17**. Recent distributions should be compatible (tested on Arch Linux, Ubuntu 19.10 and above).
@@ -23,18 +23,18 @@ ebpfpub is a generic function tracing library for Linux that supports tracepoint
     * Please note that LLVM itself must be compiled with libc++ when enabling the `EBPF_COMMON_ENABLE_LIBCPP` option, since ebfpub will directly link against the LLVM libraries.
   * The packages `llvm-devel` (for `LLVMConfig.cmake` files), `llvm-static` (for additional LLVM libraries), and `ncurses-devel` (for `libtinfo`)
 
-### Dependencies (retrieved with git)
-* [ebpf-common](https://github.com/trailofbits/ebpf-common)
-
-### Steps to Build
-
-As root, download and install the osquery-toolchain:
-
+#### Installing the osquery-toolchain
+As root:
 ```bash
 cd /tmp
 wget https://github.com/osquery/osquery-toolchain/releases/download/1.1.0/osquery-toolchain-1.1.0-x86_64.tar.xz 
 tar -xf /tmp/ebpfpub/build/osquery-toolchain-1.1.0-x86_64.tar.xz -C /opt
 ```
+
+### Dependencies (retrieved with git)
+* [ebpf-common](https://github.com/trailofbits/ebpf-common)
+
+### Steps to Build
 
 1. Obtain the source code: `git clone --recursive https://github.com/trailofbits/ebpfpub`
 2. In case the `--recursive` flag was not provided, run `git submodule update --init --recursive`

